rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,344 @@
+
+We also outline a method to compare compatibility between types, and use this in the [Rationale and Alternatives](#rationale-and-alternatives) section to explain how this replaces the need for some other versioning scheme such as semantic versioning.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Type Versions
+
+The _version number_ of a type is an incrementing positive integer that increases by 1 with each new version, starting from 1.
+
+## Type URIs
+
+The _base URI_ of a type is a _unique_ identifier for the type irrespective of its version.
+
+The _versioned URI_ of a type is made up of:
+
+```json
+"${base_uri}/v/${version_number}"
+```
+
+We opt not to constrain the format of the `base_uri` to enable flexibility in non-public use-cases or in domains with constraints that we cannot predict. Despite this, we _suggest_ that the base URIs should have a consistent format for all types hosted on your domain, and should generally be humanly readable in a way that describes your type. For example, one may pick:
+
+```json
+"http://yourdomain/namespace/property-type/type_name"
+```
+
+as the identifier (assuming that the domain decides that type names are unique across a namespace).
+
+Where it's possible to guarantee, this URI should be valid indefinitely.
+
+> ⚠️ New iterations and changes to the type must occur under new version numbers, components of the Block Protocol ecosystem will certainly implement caching and local-persistence logic, and the spec consciously makes no recommendation that these caches are updated.
+>
+> **As such, a specific version of a Type is to be treated as an immutable record**.
+
+# Reference-level explanation
+
+## What is a "version" of a Type
+
+Any published (accessible) iteration of a Type can be considered a "version" of the Type. Once a specific iteration of a type is made accessible via a URI, it should be considered immutable.
+
+- This means that any change made to the Type (for example changing the description, adding a field, etc) constitutes reason for a new version.
+- If multiple changes are made prior to making it accessible (i.e. batching the changes), then a separate version is not needed for each individual change, but they can all be 'released' together.
+
+## URI Format and Version Numbers
+
+Type URIs, and their version numbers are defined as outlined in the [Guide-level explanation](#guide-level-explanation).
+
+### Schemas
+
+1.  The `"$id"` field of a Type should be equal to the Type's _versioned URI_.
+1.  When referencing another Type using `"$ref"`, the reference should be equal to the Type's _versioned URI_.
+1.  When used as the key of a property, the _base URI_ should be used.
+
+Example:
+
+```json
+{
+  "kind": "entityType",
+  "$id": "https://blockprotocol.org/@alice/types/entity-type/book",
```

Comment:
Possibly I've misunderstood, but shouldn't this URI be versioned given the prior explanation?
```suggestion
  "$id": "https://blockprotocol.org/@alice/types/entity-type/book/v/1",
```

Reply:
You have not misunderstood, I have made a mistake 😁